core/schemas/__init__.py

```diff
@@ -19,111 +20,119 @@
+                    )
+                else:
+                    aenum.extend_enum(schema_enum, schema_type, schema_type)
+                modules.add(module_name)
+            else:
+                logger.warning(
+                    f"Observable type {schema_type} defined in <{module_name}> already exists"
+                )
+    return modules
+
+
+def _register_classes(module_name, base_module):
+    """
+    Register the classes for the schema implementation files
+
+    module_name: The module name to load
+    base_module: The base module to register the classes in (entity, indicator, observable)
+    """
+    module = importlib.import_module(module_name)
+    module_base_name = base_module.__name__.split(".")[-1]
+    schema_base_class = getattr(base_module, module_base_name.capitalize())
+    schema_type_mapping = getattr(base_module, "TYPE_MAPPING")
+    schema_types = getattr(base_module, f"{module_base_name.capitalize()}Types", None)
+    schema_enum = getattr(base_module, f"{module_base_name.capitalize()}Type")
+    for _, obj in inspect.getmembers(module, inspect.isclass):
+        if issubclass(obj, schema_base_class) and "type" in obj.model_fields:
+            obs_type = obj.model_fields["type"].default
+            logger.info(f"Registering class {obj.__name__} defining type <{obs_type}>")
+            aenum.extend_enum(schema_enum, obs_type, obs_type)
+            schema_type_mapping[obs_type] = obj
+            setattr(base_module, obj.__name__, obj)
+            if not schema_types:
+                schema_types = obj
+            else:
+                schema_types |= obj
+            setattr(base_module, f"{module_base_name.capitalize()}Types", schema_types)
+
+
+def register_classes(schema_root_type, base_module):
+    """
+    Register the classes for the schema root type
+
+    schema_root_type: The schema root type to work with (entities, indicators, observables)
+    base_module: The base module to register the classes in (entity, indicator, observable)
+    """
+    module_base_name = base_module.__name__.split(".")[-1]
+    logger.info(f"Registering {module_base_name} classes")
+    for schema_file in Path(__file__).parent.glob(f"{schema_root_type}/**/*.py"):
+        if schema_file.stem == "__init__":
             continue
-        cls = entity.TYPE_MAPPING[key]
-        if not entity.EntityTypes:
-            entity.EntityTypes = cls
-        else:
-            entity.EntityTypes |= cls
+        if schema_file.parent.stem == schema_root_type:
+            module_name = f"core.schemas.{schema_root_type}.{schema_file.stem}"
+        elif schema_file.parent.stem == "private":
+            module_name = f"core.schemas.{schema_root_type}.private.{schema_file.stem}"
+        try:
+            _register_classes(module_name, base_module)
+        except Exception:
+            logger.exception(f"Failed to register classes from {module_name}")
```

Comment:
Let's also log the error that gets thrown here

Reply:
with logger.exception, traceback is automagically appended after the message.